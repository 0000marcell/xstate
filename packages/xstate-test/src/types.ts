import { EventObject, State } from 'xstate';
export interface TestMeta<T, TContext> {
  test?: (testContext: T, state: State<TContext>) => Promise<void> | void;
  description?: string | ((state: State<TContext>) => string);
  skip?: boolean;
}
interface TestSegment<T> {
  state: State<any>;
  event: EventObject;
  description: string;
  test: (testContext: T) => Promise<void>;
  exec: (testContext: T) => Promise<void>;
}
interface TestStateResult {
  error: null | Error;
}
export interface TestSegmentResult {
  segment: TestSegment<any>;
  state: TestStateResult;
  event: {
    error: null | Error;
  };
}
interface TestPath<T> {
  weight: number;
  segments: Array<TestSegment<T>>;
  /**
   * Tests and executes each segment in `segments` sequentially, and then
   * tests the postcondition that the `state` is reached.
   */
  test: (testContext: T) => Promise<TestPathResult>;
}
export interface TestPathResult {
  segments: TestSegmentResult[];
  state: TestStateResult;
}
export interface TestPlan<T, TContext> {
  state: State<TContext>;
  paths: Array<TestPath<T>>;
  description: string;
  /**
   * Tests the postcondition that the `state` is reached.
   *
   * This should be tested after navigating any path in `paths`.
   */
  test: (testContext: T) => Promise<void>;
}
interface EventCase {
  type?: never;
  [prop: string]: any;
}
export type StatePredicate<TContext> = (state: State<TContext, any>) => boolean;
export type EventExecutor<T> = (
  testContext: T,
  event: EventObject
) => Promise<any> | void;
export interface TestModelOptions<T> {
  events: {
    [eventType: string]:
      | EventExecutor<T>
      | {
          exec?: EventExecutor<T>;
          cases?: EventCase[];
        };
  };
}
export interface TestModelCoverage {
  stateNodes: Map<string, number>;
  transitions: Map<string, Map<EventObject, number>>;
}
