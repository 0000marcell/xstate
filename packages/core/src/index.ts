import { matchesState } from './utils';
import { mapState } from './mapState';
import { StateNode } from './StateNode';
import { State } from './State';
import { Machine, createMachine } from './Machine';
import {
  raise,
  send,
  sendParent,
  sendUpdate,
  log,
  cancel,
  start,
  stop,
  assign,
  after,
  done,
  respond,
  doneInvoke
} from './actions';
import { interpret, Interpreter, spawn } from './interpreter';
import { matchState } from './match';

const actions = {
  raise,
  send,
  sendParent,
  sendUpdate,
  log,
  cancel,
  start,
  stop,
  assign,
  after,
  done,
  respond
};

export {
  Machine,
  StateNode,
  State,
  matchesState,
  mapState,
  actions,
  assign,
  send,
  sendParent,
  sendUpdate,
  interpret,
  Interpreter,
  matchState,
  spawn,
  doneInvoke,
  createMachine
};

export * from './types';
