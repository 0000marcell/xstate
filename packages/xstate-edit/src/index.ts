import { StateNode, ActionObject, Guard, InvokeDefinition } from 'xstate';
import { mapValues } from 'xstate/lib/utils';
import { getBacklinkMap } from './utils';

function getStateNodeId(stateNode: StateNode): string {
  return `#${stateNode.id}`;
}

interface StateNodeConfig {
  type: StateNode['type'];
  initial?: string;
  entry: Array<ActionObject<any, any>>;
  exit: Array<ActionObject<any, any>>;
  on: {
    [key: string]: Array<{
      target: string[];
      source: string;
      actions: Array<ActionObject<any, any>>;
      cond: Guard<any, any> | undefined;
      eventType: string;
    }>;
  };
  invoke: Array<InvokeDefinition<any, any>>;
  states: Record<string, StateNodeConfig>;
}

// derive config from machine
export function getConfig(stateNode: StateNode): StateNodeConfig {
  const config = {
    type: stateNode.type,
    initial:
      stateNode.initial === undefined ? undefined : String(stateNode.initial),
    id: stateNode.id,
    entry: stateNode.onEntry,
    exit: stateNode.onExit,
    on: mapValues(stateNode.on, tran => {
      return tran.map(t => {
        return {
          target: t.target ? t.target.map(getStateNodeId) : [],
          source: getStateNodeId(t.source),
          actions: t.actions,
          cond: t.cond,
          eventType: t.eventType
        };
      });
    }),
    invoke: stateNode.invoke,
    states: {}
  };

  Object.values(stateNode.states).forEach(sn => {
    config.states[sn.key] = getConfig(sn);
  });

  return config;
}

export function setId(
  machine: StateNode,
  id: string,
  newId: string
): StateNodeConfig {
  const stateNode = machine.getStateNodeById(id);

  stateNode.id = newId;

  return getConfig(machine);
}

export function setKey(
  machine: StateNode,
  id: string,
  newKey: string
): StateNodeConfig {
  const stateNode = machine.getStateNodeById(id);

  stateNode.key = newKey;

  return getConfig(machine);
}

export function deleteStateNode(
  machine: StateNode,
  id: string
): StateNodeConfig {
  const stateNode = machine.getStateNodeById(id);

  const backlinkMap = getBacklinkMap(machine);

  const backlinks = backlinkMap.get(stateNode) || new Set();

  backlinks.forEach(sourceNode => {
    Object.keys(sourceNode.on).forEach(key => {
      const transitions = sourceNode.on[key];
      const toDelete = [] as number[];

      transitions.forEach((transition, i) => {
        if (transition.target && transition.target.includes(stateNode)) {
          toDelete.push(i);
        }
      });

      const updatedTransitions = transitions.filter(
        (_, i) => !toDelete.includes(i)
      );

      sourceNode.on[key] = updatedTransitions;
    });
  });

  delete stateNode.parent!.states[stateNode.key];

  return getConfig(machine);
}

export function deleteTransition(
  machine: StateNode,
  id: string,
  event: string,
  index: number
): StateNodeConfig {
  const stateNode = machine.getStateNodeById(id);

  const transitions = stateNode.on[event];

  transitions.splice(index, 1);

  return getConfig(machine);
}
