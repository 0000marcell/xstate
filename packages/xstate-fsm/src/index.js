"use strict";
exports.__esModule = true;
function toArray(item) {
    return [].concat(item);
}
function FSM(fsmConfig) {
    return {
        initialState: {
            value: fsmConfig.initial,
            actions: toArray(fsmConfig.states[fsmConfig.initial].entry),
            context: fsmConfig.context
        },
        transition: function (state, event) {
            var _a = typeof state === 'string'
                ? { value: state, context: fsmConfig.context }
                : state, value = _a.value, context = _a.context;
            var eventObject = typeof event === 'string' ? { type: event } : event;
            var stateConfig = fsmConfig.states[value];
            if (stateConfig.on) {
                var transitions = [].concat(stateConfig.on[eventObject.type]);
                for (var _i = 0, transitions_1 = transitions; _i < transitions_1.length; _i++) {
                    var transition = transitions_1[_i];
                    if (transition === undefined) {
                        return { value: value, context: context, actions: [] };
                    }
                    var _b = typeof transition === 'string'
                        ? { target: transition }
                        : transition, target = _b.target, _c = _b.actions, actions = _c === void 0 ? [] : _c, _d = _b.cond, cond = _d === void 0 ? function () { return true; } : _d;
                    if (target && cond(context)) {
                        var nextStateConfig = fsmConfig.states[target];
                        var allActions = []
                            .concat(stateConfig.exit)
                            .concat(actions)
                            .concat(nextStateConfig.entry)
                            .filter(Boolean);
                        return {
                            value: target ? target : value,
                            context: context,
                            actions: allActions
                        };
                    }
                }
                return {
                    value: value,
                    context: context,
                    actions: []
                };
            }
            return { value: value, context: context, actions: [] };
        }
    };
}
exports.FSM = FSM;
