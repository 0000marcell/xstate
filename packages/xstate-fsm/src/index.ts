export type SingleOrArray<T> = T[] | T;
export interface EventObject {
  type: string;
  [key: string]: any;
}

export interface ActionObject<TContext, TEvent> {
  type: string;
  exec?: (context: TContext, event: TEvent) => void;
  [key: string]: any;
}

export type Action<TContext, TEvent extends EventObject> =
  | TEvent['type']
  | ActionObject<TContext, TEvent>
  | ((context: TContext, event: TEvent) => void);

export namespace FSM {
  export type Transition<TContext, TEvent extends EventObject> =
    | string
    | {
        target?: string;
        actions?: SingleOrArray<Action<TContext, TEvent>>;
        cond?: (context: TContext, event: TEvent) => boolean;
      };
  export interface State<TContext, TEvent> {
    value: string;
    context: TContext;
    actions: Array<ActionObject<TContext, TEvent>>;
  }
}
export interface FSMConfig<TContext, TEvent extends EventObject> {
  initial: string;
  context?: TContext;
  states: {
    [key: string]: {
      on?: {
        [event: string]: SingleOrArray<FSM.Transition<TContext, TEvent>>;
      };
      exit?: SingleOrArray<Action<TContext, TEvent>>;
      entry?: SingleOrArray<Action<TContext, TEvent>>;
    };
  };
}

function toArray<T>(item: T | T[] | undefined): T[] {
  if (item === undefined) {
    return [];
  }
  return ([] as T[]).concat(item);
}

export function assign(assignment: any): ActionObject<any, any> {
  return {
    type: 'xstate.assign',
    assignment
  };
}

export function FSM<TContext, TEvent extends EventObject = EventObject>(
  fsmConfig: FSMConfig<TContext, TEvent>
) {
  return {
    initialState: {
      value: fsmConfig.initial,
      actions: toArray(fsmConfig.states[fsmConfig.initial].entry),
      context: fsmConfig.context
    },
    transition: (
      state: string | { value: string; context: any },
      event: string | Record<string, any> & { type: string }
    ): FSM.State<TContext, TEvent> => {
      const { value, context } =
        typeof state === 'string'
          ? { value: state, context: fsmConfig.context }
          : state;
      const eventObject = (typeof event === 'string'
        ? { type: event }
        : event) as TEvent;
      const stateConfig = fsmConfig.states[value];

      if (stateConfig.on) {
        const transitions = ([] as Array<
          FSM.Transition<TContext, TEvent>
        >).concat(stateConfig.on[eventObject.type]);

        for (const transition of transitions) {
          if (transition === undefined) {
            return { value, context, actions: [] };
          }

          const { target, actions = [], cond = () => true } =
            typeof transition === 'string'
              ? { target: transition }
              : transition;

          let nextContext = context;

          if (target && cond(context, eventObject)) {
            const nextStateConfig = fsmConfig.states[target];
            const allActions = ([] as any[])
              .concat(stateConfig.exit, actions, nextStateConfig.entry)
              .filter(Boolean)
              .map<ActionObject<TContext, TEvent>>(action =>
                typeof action === 'string'
                  ? { type: action }
                  : typeof action === 'function'
                  ? {
                      type: action.name,
                      exec: action
                    }
                  : action
              )
              .filter(action => {
                if (action.type === 'xstate.assign') {
                  let tmpContext = { ...nextContext };

                  if (typeof action.assignment === 'function') {
                    tmpContext = action.assignment(nextContext, eventObject);
                  } else {
                    Object.keys(action.assignment).forEach(key => {
                      tmpContext[key] =
                        typeof action.assignment[key] === 'function'
                          ? action.assignment[key](nextContext, eventObject)
                          : action.assignment[key];
                    });
                  }

                  nextContext = tmpContext;
                  return false;
                }
                return true;
              });
            return {
              value: target ? target : value,
              context: nextContext,
              actions: allActions
            };
          }
        }
      }

      // No transitions match
      return { value, context, actions: [] };
    }
  };
}
