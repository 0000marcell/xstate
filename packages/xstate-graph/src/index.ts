import {
  getNodes,
  getEdges,
  adjacencyMap,
  simplePaths,
  getShortestPaths,
  serializeEvent,
  serializeState,
  deserializeEventString,
  deserializeStateString
} from './graph';

export {
  getNodes,
  getEdges,
  adjacencyMap as getAdjacencyMap,
  adjacencyMap,
  simplePaths as getSimplePaths,
  simplePaths,
  getShortestPaths,
  getShortestPaths as shortestPaths,
  serializeEvent,
  serializeState,
  deserializeEventString,
  deserializeStateString
};
