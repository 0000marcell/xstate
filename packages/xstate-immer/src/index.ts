import {
  EventObject,
  ActionObject,
  AssignAction,
  assign as xstateAssign
} from 'xstate';
import { produce, Draft } from 'immer';

export type ImmerAssigner<TContext, TEvent extends EventObject> = (
  context: Draft<TContext>,
  event: TEvent
) => void;

export interface ImmerAssignAction<TContext, TEvent extends EventObject>
  extends ActionObject<TContext, TEvent> {
  assignment: ImmerAssigner<TContext, TEvent>;
}

function immerAssign<TContext, TEvent extends EventObject = EventObject>(
  assignment: ImmerAssigner<TContext, TEvent>
): AssignAction<TContext, TEvent> {
  return xstateAssign((context, event) => {
    return produce(context, (draft) => void assignment(draft, event));
  });
}

export { immerAssign as assign };

export interface ImmerUpdateEventObject extends EventObject {
  input: any;
}

export interface ImmerUpdateEvent<TType extends string, TInput = any> {
  type: TType;
  input: TInput;
}

export interface ImmerUpdater<TContext, TEvent extends ImmerUpdateEventObject> {
  (input: TEvent['input']): TEvent;
  assign: ImmerAssignAction<TContext, TEvent>;
  type: TEvent['type'];
  validate?: (context: TContext, input: TEvent['input']) => boolean;
}

export function createUpdater<TContext, TEvent extends ImmerUpdateEventObject>(
  type: TEvent['type'],
  producer: (ctx: Draft<TContext>, input: TEvent['input']) => void,
  validate?: (ctx: TContext, input: TEvent['input']) => boolean
): ImmerUpdater<TContext, TEvent> {
  const updater = (input: TEvent['input']): TEvent => {
    return {
      type,
      input
    } as TEvent;
  };

  Object.assign(updater, {
    assign: immerAssign<TContext, TEvent>((ctx, event) => {
      producer(ctx, event.input);
    }),
    validate,
    type
  });

  return updater as ImmerUpdater<TContext, TEvent>;
}
