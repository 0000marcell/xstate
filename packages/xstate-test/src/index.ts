import {
  getShortestPaths,
  getSimplePaths
} from '../node_modules/@xstate/graph';
import { StateMachine, EventObject, State, StateValue } from 'xstate';

interface TestSegment<T> {
  state: State<any>;
  event: EventObject;
  test: (testContext: T) => Promise<void>;
  exec: (testContext: T) => Promise<void>;
}

interface TestPlan<T> {
  state: State<any>;
  paths: Array<{
    weight: number;
    segments: Array<TestSegment<T>>;
  }>;
  test: (testContext: T) => Promise<void>;
}

interface EventSample {
  type: never;
  [prop: string]: any;
}

type StatePredicate<TContext> = (state: State<TContext, any>) => boolean;

interface TestModelOptions<T> {
  events: {
    [eventType: string]: {
      exec: (testContext: T, event: EventObject) => Promise<any>;
      samples?: EventSample[];
    };
  };
}

interface TestModelCoverage {
  stateNodes: Map<string, number>;
  transitions: Map<string, Map<EventObject, number>>;
}

export class TestModel<T, TContext> {
  public coverage: TestModelCoverage = {
    stateNodes: new Map(),
    transitions: new Map()
  };

  constructor(
    public machine: StateMachine<TContext, any, any>,
    public options: TestModelOptions<T>
  ) {}

  public getShortestPaths(
    options?: Parameters<typeof getShortestPaths>[1]
  ): Array<TestPlan<T>> {
    const shortestPaths = getShortestPaths(this.machine, {
      ...options,
      events: getEventSamples<T>(this.options.events)
    });

    return Object.keys(shortestPaths).map(key => {
      const testPlan = shortestPaths[key];

      return {
        ...testPlan,
        test: testContext => this.test(testPlan.state, testContext),
        paths: [
          {
            weight: testPlan.weight || 0,
            segments: testPlan.path.map(segment => {
              return {
                ...segment,
                test: testContext => this.test(segment.state, testContext),
                exec: testContext => this.exec(segment.event, testContext)
              };
            })
          }
        ]
      };
    });
  }

  public getShortestPathsTo(
    stateValue: StateValue | StatePredicate<TContext>
  ): Array<TestPlan<T>> {
    let minWeight = Infinity;
    let shortestPlans: Array<TestPlan<T>> = [];

    const plans = this.filterPathsTo(stateValue, this.getShortestPaths());

    for (const plan of plans) {
      const currWeight = plan.paths[0].weight;
      if (currWeight < minWeight) {
        minWeight = currWeight;
        shortestPlans = [plan];
      } else if (currWeight === minWeight) {
        shortestPlans.push(plan);
      }
    }

    return shortestPlans;
  }

  private filterPathsTo(
    stateValue: StateValue | StatePredicate<TContext>,
    testPlans: Array<TestPlan<T>>
  ): Array<TestPlan<T>> {
    const predicate =
      typeof stateValue === 'function'
        ? plan => stateValue(plan.state)
        : plan => plan.state.matches(stateValue);
    return testPlans.filter(predicate);
  }

  public getSimplePaths(
    options?: Parameters<typeof getSimplePaths>[1]
  ): Array<TestPlan<T>> {
    const simplePaths = getSimplePaths(this.machine, {
      ...options,
      events: getEventSamples(this.options.events)
    });

    return Object.keys(simplePaths).map(key => {
      const testPlan = simplePaths[key];

      return {
        ...testPlan,
        test: testContext => this.test(testPlan.state, testContext),
        paths: testPlan.paths.map(segments => {
          return {
            weight: 0,
            segments: segments.map(segment => {
              return {
                ...segment,
                test: testContext => this.test(segment.state, testContext),
                exec: testContext => this.exec(segment.event, testContext)
              };
            })
          };
        })
      };
    });
  }

  public getSimplePathsTo(
    stateValue: StateValue | StatePredicate<TContext>
  ): Array<TestPlan<T>> {
    return this.filterPathsTo(stateValue, this.getSimplePaths());
  }

  public async test(state: State<any, any>, testContext: T) {
    for (const id of Object.keys(state.meta)) {
      const stateNodeMeta = state.meta[id];
      if (typeof stateNodeMeta.test === 'function') {
        // console.log('fn');
        this.coverage.stateNodes.set(
          id,
          (this.coverage.stateNodes.get(id) || 0) + 1
        );

        // console.log([...this.coverage.stateNodes.keys()]);

        await stateNodeMeta.test(testContext);
      }
    }
  }

  public async exec(event: EventObject, testContext: T) {
    const testEvent = this.options.events[event.type];

    if (!testEvent) {
      throw new Error(`no event configured for ${event.type}`);
    }

    await testEvent.exec(testContext, event);
  }

  public getCoverage(): { stateNodes: Record<string, number> } {
    const coverage = {
      stateNodes: {}
    };

    for (const key of this.coverage.stateNodes.keys()) {
      coverage.stateNodes[key] = this.coverage.stateNodes.get(key);
    }

    return coverage;
  }
}

function getEventSamples<T>(eventsOptions: TestModelOptions<T>['events']) {
  const result = {};

  Object.keys(eventsOptions).forEach(key => {
    const eventOptions = eventsOptions[key];
    result[key] = eventOptions.samples
      ? eventOptions.samples.map(sample => ({
          type: key,
          ...sample
        }))
      : [
          {
            type: key
          }
        ];
  });

  return result;
}

export function createModel<TestContext, TContext = any>(
  machine: StateMachine<TContext, any, any>,
  options: TestModelOptions<TestContext>
): TestModel<TestContext, TContext> {
  return new TestModel<TestContext, TContext>(machine, options);
}
