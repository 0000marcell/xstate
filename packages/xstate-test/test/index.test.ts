// nothing yet
import { createModel } from '../src';
import { Machine, assign } from 'xstate';

const dieHardMachine = Machine<{ three: number; five: number }>(
  {
    id: 'dieHard',
    initial: 'pending',
    context: { three: 0, five: 0 },
    states: {
      pending: {
        on: {
          '': {
            target: 'success',
            cond: 'weHave4Gallons'
          },
          POUR_3_TO_5: {
            actions: 'pour3to5'
          },
          POUR_5_TO_3: {
            actions: 'pour5to3'
          },
          FILL_3: {
            actions: 'fill3'
          },
          FILL_5: {
            actions: 'fill5'
          },
          EMPTY_3: {
            actions: 'empty3'
          },
          EMPTY_5: {
            actions: 'empty5'
          }
        },
        meta: {
          description: state => {
            return `pending with (${state.context.three}, ${state.context.five})`;
          },
          test: async ({ jugs }, state) => {
            expect(jugs.five).not.toEqual(4);
            expect(jugs.three).toEqual(state.context.three);
            expect(jugs.five).toEqual(state.context.five);
          }
        }
      },
      success: {
        type: 'final',
        meta: {
          description: 'succeeds with 4 gallons',
          test: async ({ jugs }) => {
            expect(jugs.five).toEqual(4);
          }
        }
      }
    }
  },
  {
    actions: {
      pour3to5: assign(ctx => {
        const poured = Math.min(5 - ctx.five, ctx.three);

        return {
          three: ctx.three - poured,
          five: ctx.five + poured
        };
      }),
      pour5to3: assign(ctx => {
        const poured = Math.min(3 - ctx.three, ctx.five);

        const res = {
          three: ctx.three + poured,
          five: ctx.five - poured
        };

        return res;
      }),
      fill3: assign({ three: 3 }),
      fill5: assign({ five: 5 }),
      empty3: assign({ three: 0 }),
      empty5: assign({ five: 0 })
    },
    guards: {
      weHave4Gallons: ctx => ctx.five === 4
    }
  }
);

class Jugs {
  public three = 0;
  public five = 0;

  public fillThree() {
    this.three = 3;
  }
  public fillFive() {
    this.five = 5;
  }
  public emptyThree() {
    this.three = 0;
  }
  public emptyFive() {
    this.five = 0;
  }
  public transferThree() {
    const poured = Math.min(5 - this.five, this.three);

    this.three = this.three - poured;
    this.five = this.five + poured;
  }
  public transferFive() {
    const poured = Math.min(3 - this.three, this.five);

    this.three = this.three + poured;
    this.five = this.five - poured;
  }
}

const dieHardModel = createModel<{ jugs: Jugs }>(dieHardMachine, {
  events: {
    POUR_3_TO_5: {
      exec: async ({ jugs }) => {
        await jugs.transferThree();
      }
    },
    POUR_5_TO_3: {
      exec: async ({ jugs }) => {
        await jugs.transferFive();
      }
    },
    EMPTY_3: {
      exec: async ({ jugs }) => {
        await jugs.emptyThree();
      }
    },
    EMPTY_5: {
      exec: async ({ jugs }) => {
        await jugs.emptyFive();
      }
    },
    FILL_3: {
      exec: async ({ jugs }) => {
        await jugs.fillThree();
      }
    },
    FILL_5: {
      exec: async ({ jugs }) => {
        await jugs.fillFive();
      }
    }
  }
});

describe('testing a model (shortestPathsTo)', () => {
  dieHardModel
    .getShortestPathsTo('success') // ...
    .forEach((plan, planIndex) => {
      describe(`reaches state ${JSON.stringify(
        plan.state.value
      )} (${planIndex})`, () => {
        const testJugs = new Jugs();

        plan.paths.forEach((path, pathIndex) => {
          describe(`path ${pathIndex}`, () => {
            path.segments.forEach(segment => {
              it(segment.description, async () => {
                await segment.test({ jugs: testJugs });
              });

              it(`executes ${JSON.stringify(segment.event)}`, async () => {
                await segment.exec({ jugs: testJugs });
              });
            });

            it(plan.description, async () => {
              await plan.test({ jugs: testJugs });
            });
          });
        });
      });
    });
});

describe('testing a model (simplePathsTo)', () => {
  dieHardModel
    .getSimplePathsTo('success') // ...
    .forEach(plan => {
      describe(`reaches state ${JSON.stringify(
        plan.state.value
      )} (${JSON.stringify(plan.state.context)})`, () => {
        plan.paths.forEach((path, pathIndex) => {
          describe(`path ${pathIndex}`, () => {
            const testJugs = new Jugs();

            path.segments.forEach(segment => {
              it(`goes to ${JSON.stringify(
                segment.state.value
              )} ${JSON.stringify(segment.state.context)}`, async () => {
                await segment.test({ jugs: testJugs });
              });

              it(`executes ${JSON.stringify(segment.event)}`, async () => {
                await segment.exec({ jugs: testJugs });
              });
            });

            it('reaches the final state', async () => {
              await plan.test({ jugs: testJugs });
            });
          });
        });
      });
    });
});

describe('path.test()', () => {
  const plans = dieHardModel.getSimplePathsTo(state => {
    return state.matches('success') && state.context.three === 0;
  });

  // ...
  plans.forEach(plan => {
    describe(`reaches state ${JSON.stringify(
      plan.state.value
    )} (${JSON.stringify(plan.state.context)})`, () => {
      plan.paths.forEach((path, pathIndex) => {
        describe(`path ${pathIndex}`, () => {
          const testJugs = new Jugs();

          it(`reaches the target state`, () => {
            return path.test({ jugs: testJugs });
          });
        });
      });
    });
  });
});

describe('coverage', () => {
  it('reports state node coverage', () => {
    const coverage = dieHardModel.getCoverage();

    expect(coverage.stateNodes['dieHard.pending']).toBeGreaterThan(0);
    expect(coverage.stateNodes['dieHard.success']).toBeGreaterThan(0);
  });
});
