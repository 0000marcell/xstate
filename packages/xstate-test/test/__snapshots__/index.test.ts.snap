// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error path trace should return trace for failed state should show an error path trace: error path trace 1`] = `
"test error
Path:
	State: [92m\\"first\\" [39m
	Event: [32m{\\"type\\":\\"NEXT\\"}[39m

	State: [92m\\"second\\" [39m
	Event: [32m{\\"type\\":\\"NEXT\\"}[39m

	State: [31m\\"third\\" [39m"
`;
